doc.go

```diff
@@ -0,0 +1,113 @@
+// This source file is part of the EdgeDB open source project.
+//
+// Copyright 2020-present EdgeDB Inc. and the EdgeDB authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package edgedb is the official go EdgeDB driver. https://edgedb.com
+//
+// Typical usage looks like this:
+//
+//   opts := edgedb.Options{
+//       MinConns: 1,
+//       MaxConns: 4,
+//   }
+//
+//   ctx := context.Background()
+//   pool, err := edgedb.ConnectDSN(ctx, "my_instance", opts)
+//   if err != nil {
+//       log.Fatal(err)
+//   }
+//   defer pool.Close()
+//
+//   var (
+//       age int64 = 21
+//       users []struct{
+//           ID edgedb.UUID `edgedb:"id"`
+//           Name string    `edgedb:"name"`
+//       }
+//   )
+//
+//   query := "SELECT User{name} WHERE .age = <int64>$0"
+//   err = pool.Query(ctx, query, &users, age)
+//   ...
+//
+// You can also connect to a database using a url:
+//
+//   url := "edgedb://edgedb@localhost/edgedb"
+//   pool, err := edgedb.ConnectDSN(ctx, url, opts)
+//
+// Or you can use Option fields.
+//
+//   opts := edgedb.Options{
+//       Database: "edgedb",
+//       User:     "edgedb",
+//       MinConns: 1,
+//       MaxConns: 4,
+//   }
+//
+//   pool, err := edgedb.Connect(ctx, opts)
+//
+// Pooling
+//
+// Most use cases will benefit from the concurrency safe pool implementation
+// returned from Connect() and ConnectDSN(). Pool.Acquire(), ConnectOne() and
+// ConnectOneDSN() will give you access to a single connection,
+// but be aware that connections will reconnect automatically on network errors
+// so use TryTx() and Retry() when you want to run queries in a transaction.
```

Comment:
TryTx feels weird in this sentence since it wouldn't try to reconnect on network errors.

Reply:
Exactly. If you try to roll your own transactions with our driver you can burn yourself if there is a reconnect.

```go
con.Execute(ctx, "START TRANSACTION")
con.Query(...) // connections fails
con.Query(...) // new connection i.e. different transaction
con.Execute(ctx, "COMMIT")
```

The only safe way to execute multiple queries in the same transaction is with TryTx() or Retry().

Reply:
> you can burn yourself if there is a reconnect.

Hm, we shouldn't let that happen and this needs to be fixed. If the connection in a transaction that was manually started, we should error out without trying to reconnect.

Reply:
@tailhook Paul, we need to make sure we don't make the same mistake in edgedb-js/edgedb-python

Reply:
Adding further to the subject: Retry() should reconnect on network errors and try the transaction again. TryTx() should just error out if an error happened.

Reply:
issue created here https://github.com/edgedb/edgedb-go/issues/96 for edgedb-go